File: python/src/azoth/eos/reference/pr_danesh_alpha.py
"""``eos.pr_danesh_alpha`` - the Danesh alpha function.

Spec: ``specs/calcs/eos/pr_danesh_alpha.toml``, which carries the provenance and why the
1.21 factor damps the attraction past the critical temperature.
"""

from __future__ import annotations

import math

from azoth._registry_gen import spec as _spec_for
from azoth.core.range import apply_checks, checks_for
from azoth.core.result import PrDaneshAlphaResult
from azoth.core.warnings import Warning

CALC_ID = "eos.pr_danesh_alpha"


def pr_danesh_alpha(omega: float, Tr: float) -> PrDaneshAlphaResult:
    """The Danesh alpha function for a pure component.

    Args:
        omega: the acentric factor.
        Tr: the reduced temperature.

    Returns:
        The temperature-dependent alpha function.

    Raises:
        OutOfRangeError: if ``Tr`` is not positive.

    Example:
        >>> pr_danesh_alpha(0.152, 0.7).alpha
        1.2066270990034567
    """
    spec = _spec_for(CALC_ID)
    checks = checks_for(spec)
    warnings: list[Warning] = []

    values = {"omega": omega, "Tr": Tr}
    apply_checks(checks.on_input, values.get, warnings)

    m = 0.37464 + 1.54226 * omega - 0.26992 * omega * omega
    m_mod = 1.21 * m if Tr > 1.0 else m
    one_minus = 1.0 - math.sqrt(Tr)
    alpha = (1.0 + m_mod * one_minus) * (1.0 + m_mod * one_minus)

    apply_checks(checks.derived, lambda name: alpha if name == "alpha" else None, warnings)

    return PrDaneshAlphaResult(alpha=alpha, warnings=tuple(warnings))

File: crates/azoth-eos/src/pr_danesh_alpha.rs
//! `eos.pr_danesh_alpha` - the Danesh alpha function.
//!
//! ```text
//! m = 0.37464 + 1.54226*omega - 0.26992*omega**2
//! m_mod = 1.21*m above Tr = 1 else m
//! alpha = (1 + m_mod*(1 - sqrt(Tr)))**2
//! ```
//!
//! Spec: `specs/calcs/eos/pr_danesh_alpha.toml`, which carries the provenance and why the
//! 1.21 factor damps the attraction past the critical temperature.

use azoth_core::{Result, apply_checks};

use crate::results::PrDaneshAlphaResult;
use crate::spec_gen;

/// The Danesh alpha function for a pure component.
///
/// `omega` is the acentric factor; `Tr` the reduced temperature. Above `Tr = 1` the
/// Peng-Robinson Soave coefficient is multiplied by 1.21, which damps the attraction
/// into the supercritical region.
///
/// # Errors
/// * [`azoth_core::AzothError::OutOfRange`] if `Tr <= 0` - the formula takes a square
///   root of `Tr`.
///
/// # Example
/// ```
/// use azoth_eos::pr_danesh_alpha;
///
/// let r = pr_danesh_alpha(0.152, 0.7)?;
/// assert!((r.alpha - 1.2066270990034567).abs() < 1e-15);
/// # Ok::<(), azoth_core::AzothError>(())
/// ```
#[allow(non_snake_case)] // `Tr` is the symbol in the published equation
pub fn pr_danesh_alpha(omega: f64, Tr: f64) -> Result<PrDaneshAlphaResult> {
    let spec = &spec_gen::PR_DANESH_ALPHA_SPEC;
    let mut warnings = Vec::new();

    apply_checks(
        spec.input_checks(),
        |quantity| match quantity {
            "omega" => Some(omega),
            "Tr" => Some(Tr),
            _ => None,
        },
        &mut warnings,
    )?;

    let m = 0.37464 + 1.54226 * omega - 0.26992 * omega * omega;
    let m_mod = if Tr > 1.0 { 1.21 * m } else { m };
    let alpha = (1.0 + m_mod * (1.0 - Tr.sqrt())) * (1.0 + m_mod * (1.0 - Tr.sqrt()));

    apply_checks(
        spec.derived_checks(),
        |name| (name == "alpha").then_some(alpha),
        &mut warnings,
    )?;

    Ok(PrDaneshAlphaResult { alpha, warnings })
}
